### PR TITLE
feat: Propose threshold change

### PR DESCRIPTION
**Pull request by 0xnandagopan**
closes [issue #63](https://github.com/Spherre-Labs/spherre/issues/63)

### Overview
This PR introduces a new feature that allows members with the `PROPOSER` role to submit a transaction proposal for updating the treasury's `threshold` value (of type `u64`). This threshold defines the minimum number of votes required for a transaction to be approved in the multisig setup.

The implementation enforces strict validation rules to ensure that the new threshold is reasonable and does not compromise system integrity.

---

#### + `ichange_threshold_tx.cairo`
- Starknet interface defining allowed functions for threshold change proposals.

#### + `change_threshold_transaction.cairo`
- Core Starknet component implementing the logic to propose a threshold change.
- Includes validation and integrates with the `account_data` component.

#### + `mock_account_data.cairo`
- Mock contract extended to incorporate threshold change proposal logic.
- Public versions of component interfaces are exposed via the mock ABI.

#### + `test_change_threshold_transaction.cairo`
- Comprehensive test suite validating success and major edge cases:
  - Proposal by user without the required role.
  - Setting the threshold to `0`.
  - Attempting a proposal with no actual change in threshold value.
  - Setting a threshold value higher than the total number of members.
  - Accessing invalid or non-existent proposals.
  - Proposal attempt while the functionality is paused.
- Tests are executed on the `mock_account_data` contract.

#### + `errors.cairo`
- Defines new feature-specific error values.

#### + `lib.cairo`
- Registers the new module within the project structure.

---

### Next Steps
Implement the execution logic to process approved threshold change proposals, ensuring alignment with the governing infrastructure agreements.

---

###  / nandagopan .

[twitter](https://x.com/0xnandagopan) | [telegram](https://t.me/Oxnandagopan)

## Summary by CodeRabbit

- **New Features**
  - Introduced the ability to propose, store, and retrieve threshold change transactions with appropriate permission checks and event notifications.
  - Added support for listing all threshold change transactions.
- **Bug Fixes**
  - Improved error handling for threshold validation and transaction retrieval with new descriptive error messages.
- **Tests**
  - Added comprehensive tests covering successful and failed threshold change proposals, event emission, and retrieval scenarios.

### CODE REVIEW

File: src/tests/actions/test_change_threshold_transaction.cairo [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    // Propose two threshold change transactions
+    let tx_id1 = mock_contract.propose_threshold_change_transaction_pub(4);
+    let tx_id2 = mock_contract.propose_threshold_change_transaction_pub(2);
+    stop_cheat_caller_address(mock_contract.contract_address);
+
+    // Retrieve all transactions
+    let transactions = mock_contract.get_all_threshold_change_transactions_pub();
+    assert(transactions.len() == 2, 'Incorrect transaction count');
+    assert(*transactions.at(0).new_threshold == 4, 'Incorrect first threshold');
+    assert(*transactions.at(1).new_threshold == 2, 'Incorrect second threshold');
```
[the remark this answers is not in the document]
Comment (codebestia) [reply src/tests/actions/test_change_threshold_transaction.cairo:269]: Please resolve this

[the remark this answers is not in the document]
Comment (0xnandagopan) [reply src/tests/actions/test_change_threshold_transaction.cairo:269]: > GM @0xnandagopan Please resolve this

Hi @codebestia , on it 👍

**Review by codebestia — APPROVED** [review]
LGTM!
Thank you for your contribution @0xnandagopan
